src/llm/llmnoderesources.cpp

```diff
@@ -38,20 +38,90 @@
+        auto locals = py::dict("models_path"_a = nodeResources->modelsPath);
+        py::exec(R"(
+            # Following the logic from:
+            # https://github.com/huggingface/transformers/blob/25245ec26dc29bcf6102e1b4ddd0dfd02e720cf5/src/transformers/tokenization_utils_base.py#L1837
+
+            global json
+            import json
+            from pathlib import Path
+
+            global jinja2
+            import jinja2
+            from jinja2.sandbox import ImmutableSandboxedEnvironment
+
+            def raise_exception(message):
+                raise jinja2.exceptions.TemplateError(message)
+
+
+            # Default chat template accepts only single message and outputs only it's 'content'
+            # effectively turning it into a regular prompt. 
+            default_chat_template = "{% if messages|length > 1 %} {{ raise_exception('This servable accepts only single message requests') }}{% endif %}{{ messages[0]['content'] }}"
+
+            bos_token = ""
+            eos_token = ""
+            chat_template = default_chat_template
+
+            template = None
+
+            # Try to read template from template.jinja file
+            jinja_file = Path(models_path + "/template.jinja")
+            if jinja_file.is_file():
+                template_loader = jinja2.FileSystemLoader(searchpath=models_path)
+                jinja_env = ImmutableSandboxedEnvironment(trim_blocks=True, lstrip_blocks=True, loader=template_loader)
+                jinja_env.policies["json.dumps_kwargs"]["ensure_ascii"] = False
+                jinja_env.globals["raise_exception"] = raise_exception
+                template = jinja_env.get_template("template.jinja")
+
+            # Try to read data from tokenizer_config.json
+            tokenizer_config_file = Path(models_path + "/tokenizer_config.json")
+            if tokenizer_config_file.is_file():
+                f = open(models_path + "/tokenizer_config.json")
+                data = json.load(f)
+                bos_token = data.get("bos_token", "")
+                eos_token = data.get("eos_token", "")
+                chat_template = data.get("chat_template", default_chat_template)
+
+            if template is None:
+                jinja_env = ImmutableSandboxedEnvironment(trim_blocks=True, lstrip_blocks=True)
+                jinja_env.policies["json.dumps_kwargs"]["ensure_ascii"] = False
+                jinja_env.globals["raise_exception"] = raise_exception
+                template = jinja_env.from_string(chat_template)
+        )",
+            py::globals(), locals);
+
+        nodeResources->textProcessor.bosToken = locals["bos_token"].cast<std::string>();
+        nodeResources->textProcessor.eosToken = locals["eos_token"].cast<std::string>();
+        nodeResources->textProcessor.chatTemplate = std::make_unique<PyObjectWrapper<py::object>>(locals["template"]);
+    } catch (const pybind11::error_already_set& e) {
+        SPDLOG_INFO(CHAT_TEMPLATE_WARNING_MESSAGE);
```

Comment:
Why shouldnt we fail the resource loading phase in case the template cannot be read? @mzegla

Reply:
Because this definition could potentially be used by the graph that doesn't require chat template (completions for example).